### PR TITLE
Add GPT-OSS 120B TP8-DP4-EP32 benchmark on Galaxy

### PR DESCRIPTION
**Pull request by odjuricicTT**
## Summary
- Add GPT-OSS 120B benchmark with TP=8, DP=4, EP=32 on a 4x8 Galaxy mesh (batch size 128)
- Extend benchmark framework with `input_output_sharding_spec` and `kv_cache_sharding_spec` parameters for flexible input/KV cache sharding
- Update `LLMSamplingWrapper` to produce sharded and replicated copies of `next_token_ids` to avoid recompilation during decode
- Add the new test to the perf-bench CI matrix

## What's changed
- **`tests/benchmark/test_llms.py`**: New test `test_gpt_oss_120b_tp_dp_galaxy_batch_size_128` with a 4x8 mesh config and MoE sharding specs — attention weights TP=8, expert weights EP=32, inputs DP=4
- **`tests/benchmark/benchmarks/llm_benchmark.py`**: New `input_output_sharding_spec` and `kv_cache_sharding_spec` parameters to allow per-test control of input and KV cache sharding (previously hardcoded)
- **`tests/benchmark/llm_utils/decode_utils.py`**: `LLMSamplingWrapper` now returns both sharded and replicated `next_token_ids` — sharded version feeds the next decode step, replicated version transfers to CPU without triggering an extra all-gather graph
- **`.github/workflows/perf-bench-matrix.json`**: Added the new test to the CI matrix on `galaxy-wh-6u`

## Perf benchmark run
https://github.com/tenstorrent/tt-xla/actions/runs/24247948813

## Known Issues
- PCC is bad and needs to be debugged — tracked in #4206

### CODE REVIEW

**Comment by odjuricicTT** [comment]
## Benchmark Runs

**Perf benchmarks** (all HW configs: n150, p150, n300-llmbox, galaxy-wh-6u — non-shared runners):
https://github.com/tenstorrent/tt-xla/actions/runs/24280697969

**Accuracy benchmarks** (n150, n300-llmbox, galaxy-wh-6u):
https://github.com/tenstorrent/tt-xla/actions/runs/24280698256

**Review by vvukomanTT — APPROVED** [review]

**Review by nsumrakTT — APPROVED** [review]

**Review by vkovacevicTT — APPROVED** [review]

File: tests/benchmark/benchmarks/llm_benchmark.py [hunk 1]
```diff
@@ -376,9 +378,15 @@ def benchmark_llm_torch_xla(
                 xs.mark_sharding(tensor, mesh, shard_spec)
 
         # Also shard KV cache tensors created in input_args
+        # This is hardcoded for all TP models, and should be moded to tt-forge-models.
+        kv_spec = kv_cache_sharding_spec or (None, "model", None, None)
```
Comment (acicovicTT) [inline tests/benchmark/benchmarks/llm_benchmark.py:384]: Can you just make `kv_cache_sharding_spec` param have the default value of `(None, "model", None, None)`?

File: tests/benchmark/test_llms.py [hunk 2]
```diff
@@ -1621,3 +1625,82 @@ def test_gpt_oss_120b_tp_galaxy_batch_size_64(
...
+
+def _moe_throughput_galaxy_shard_spec_fn(model_loader, model):
+    """Sharding specs for MoE models optimized for throughput on 4x8 galaxy mesh.
+    TP - 8 : DP - 4 : EP - 32
+    Inputs are sharded on the batch axis DP - 4. One tile per device so batch 128 should be used.
+    Attention weights are sharded on model axis TP - 8 and replicated along the batch axis.
+    Expert weights are sharded across both model and batch axes EP - 32.
+    """
+
+    batch_axis = None
```
Comment (acicovicTT) [inline tests/benchmark/test_llms.py]: Why not use `None` everywhere? It's either `(None, None)` or `("model", None)` implying batch. Also `(batch_axis,)` was not obvious to me while `(None,)` would be.

File: tests/benchmark/llm_utils/decode_utils.py [hunk 3]
```diff
@@ -52,12 +61,38 @@ def forward(self, input_ids, past_key_values, cache_position, use_cache=True):
...
-        last_token_logits = logits[:, -1]
-        next_token_ids = last_token_logits.argmax(dim=-1, keepdim=True)
+        # Only take logits for last token in prefill.
+        # This is a noop for decode.
+        logits_last = logits[:, -1]
+        next_token_ids = logits_last.argmax(dim=-1, keepdim=True)
+        next_token_ids_replicated = next_token_ids
+        if self.mesh and self.output_sharding_spec:
+            # Create two versions of next_token_ids, sharded and replicated.
+            # The sharded version is used as input for the next decode step. Passing the replicated version as the input created a differenct graph and trigers recompilation.
```
Comment (acicovicTT) [inline tests/benchmark/llm_utils/decode_utils.py]: Nit, typos: different, triggers and creates (next line)

**Review by acicovicTT — APPROVED** [review]
LGTM, left some suggestions

**Comment by odjuricicTT** [comment]
Galaxy perf benchmark run for GPT-OSS 120B TP8-DP4-EP32: https://github.com/tenstorrent/tt-xla/actions/runs/24406714242

File: tests/benchmark/test_llms.py (hunk 2 above)
Comment (odjuricicTT) [reply tests/benchmark/test_llms.py]: Was just a stale change. Thanks for pointing out! Changed.

File: tests/benchmark/benchmarks/llm_benchmark.py (hunk 1 above)
Comment (odjuricicTT) [reply tests/benchmark/benchmarks/llm_benchmark.py:384]: Will leave here for now as it's more explicit that this is hardcoded (which it should not be).

Opened an issue to move this to tt-forge-models which should be the long term solution.
